Extensions/render/image.lisp

```diff
@@ -130,47 +130,127 @@
+                              nil)
+                             ((let ((value (aref stencil-array stencil-y stencil-x)))
+                                (if (= #xff value)
+                                    nil
+                                    value))))))))
+           (update-ink (i j)
+             (setf ink (clime:design-ink design i j))
+             (when (and (eq old-ink ink) (eql old-alpha alpha))
+               (return-from update-ink))
+
+             (cond ((typep ink 'standard-flipping-ink)
+                    (let-rgba ((r.fg g.fg b.fg a.fg) (let ((d1 (slot-value ink 'climi::design1))
+                                                           (d2 (slot-value ink 'climi::design2)))
+                                                       (logior (logxor (climi::%rgba-value d1)
+                                                                       (climi::%rgba-value d2))
+                                                               #xff)))
+                      (setf source-r r.fg
+                            source-g g.fg
+                            source-b b.fg)
+                      (cond (alpha
+                             (setf source-a (%octet-mult a.fg alpha)
+                                   mode     :flipping/blend))
+                            (t
+                             (setf source-a    #x00
+                                   source-rgba (%vals->rgba source-r source-g source-b #xff)
+                                   mode        :flipping)))))
+                   ((not alpha)
+                    (let ((ink-rgba (climi::%rgba-value ink)))
+                      (if (= #xff (logand #xff ink-rgba))
+                          (setf source-rgba ink-rgba
+                                mode        :copy)
+                          (let-rgba ((r g b a) ink-rgba)
+                            (setf source-r r
+                                  source-g g
+                                  source-b b
+                                  source-a a
+                                  mode     :blend)))))
+                   (t
+                    (let-rgba ((r.fg g.fg b.fg a.fg) (climi::%rgba-value ink))
+                      (setf source-r r.fg
+                            source-g g.fg
+                            source-b b.fg
+                            source-a (%octet-mult a.fg alpha)
+                            mode     (case source-a
+                                       (0
+                                        nil)
+                                       (255
+                                        (setf source-rgba (climi::%rgba-value ink))
+                                        :copy)
+                                       (t
+                                        :blend))))))))
+      (do-regions ((src-j j y y y2)
+                   (src-i i x x x2))
+        (when (or (null clip-region)
+                  (region-contains-position-p clip-region src-i src-j))
```

Comment:
could we lift up check for null clip-region and duplicate the do-regions form (one with this one and another without it)?
```
(if clip-region
   (do-regions () (when contains-position foo))
   (do-regiuons () foo))
```